File: app/validators.py
# app/validators.py
import re
from decimal import Decimal
from typing import Tuple, Dict, Any
from datetime import datetime, timedelta
import logging  # NOVO: Para logs de debug

logger = logging.getLogger(__name__)

# ============================================================================
# VALIDADORES DE CNPJ
# ============================================================================
def validar_cnpj(cnpj: str) -> Tuple[bool, str]:
    """
    Valida o CNPJ usando o algoritmo correto de dígito verificador.
    Args:
        cnpj: String com o CNPJ (com ou sem formatação)
    Returns:
        Tuple[bool, str]: (é_válido, mensagem_erro)
    Exemplo:
        >>> validar_cnpj("11.222.333/0001-81")
        (True, "")
    """
    # Limpeza robusta: remove não-dígitos e espaços
    cnpj_limpo = re.sub(r'\D', '', cnpj.strip())
    
    # Log para debug (veja no console)
    logger.info(f"Validando CNPJ limpo: {cnpj_limpo}")
    
    # Verificar tamanho
    if len(cnpj_limpo) != 14:
        logger.warning(f"Tamanho inválido: {len(cnpj_limpo)} dígitos")
        return False, "CNPJ deve ter 14 dígitos."
    
    # Verificar se todos os dígitos são iguais (CNPJ inválido)
    if cnpj_limpo == cnpj_limpo[0] * 14:
        logger.warning("Todos dígitos iguais")
        return False, "CNPJ inválido (todos os dígitos iguais)."
    
    # ✅ ALGORITMO CORRETO - Primeiro dígito verificador (DV1)
    soma = 0
    peso = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
    for i in range(12):
        soma += int(cnpj_limpo[i]) * peso[i]
    
    resto = soma % 11
    primeiro_digito = 0 if resto < 2 else 11 - resto
    
    logger.info(f"DV1 calculado: {primeiro_digito}, real: {cnpj_limpo[12]}")
    
    if int(cnpj_limpo[12]) != primeiro_digito:
        return False, "CNPJ inválido. Dígito verificador 1 incorreto."
    
    # ✅ ALGORITMO CORRETO - Segundo dígito verificador (DV2)
    soma = 0
    peso = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
    for i in range(13):
        soma += int(cnpj_limpo[i]) * peso[i]
    
    resto = soma % 11
    segundo_digito = 0 if resto < 2 else 11 - resto
    
    logger.info(f"DV2 calculado: {segundo_digito}, real: {cnpj_limpo[13]}")
    
    if int(cnpj_limpo[13]) != segundo_digito:
        return False, "CNPJ inválido. Dígito verificador 2 incorreto."
    
    logger.info("CNPJ validado com sucesso")
    return True, ""

# ============================================================================
# VALIDADORES DE EMAIL
# ============================================================================

def validar_email(email: str) -> Tuple[bool, str]:
    """
    Valida o formato do email.
    
    Args:
        email: String com o email
    
    Returns:
        Tuple[bool, str]: (é_válido, mensagem_erro)
    """
    
    padrao = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    
    if not re.match(padrao, email):
        return False, "Email inválido. Verifique o formato."
    
    return True, ""

# ============================================================================
# VALIDADORES DE TELEFONE
# ============================================================================

def validar_telefone(telefone: str) -> Tuple[bool, str]:
    """
    Valida o formato do telefone.
    Aceita formatos: (XX) XXXX-XXXX ou XXXXXXXXXX
    
    Args:
        telefone: String com o telefone
    
    Returns:
        Tuple[bool, str]: (é_válido, mensagem_erro)
    
    Exemplo:
        >>> validar_telefone("(67) 3333-3333")
        (True, "")
        >>> validar_telefone("6733333333")
        (True, "")
    """
    # Remover formatação
    numeros = re.sub(r'\D', '', telefone)
    
    # Verificar se tem 10 dígitos (2 de DDD + 8 dígitos do telefone)
    if len(numeros) != 10:
        return False, "Telefone inválido. Use (XX) XXXX-XXXX ou XXXXXXXXXX."
    
    return True, ""


def validar_whatsapp(whatsapp: str) -> Tuple[bool, str]:
    """
    Valida o formato do WhatsApp.
    Aceita formatos com 10 ou 11 dígitos (celulares ou fixos).
    
    Args:
        whatsapp: String com o WhatsApp
    
    Returns:
        Tuple[bool, str]: (é_válido, mensagem_erro)
    """
    # Remover formatação (deixa só os números)
    numeros = re.sub(r'\D', '', whatsapp)
    
    # Verificar se tem 10 ou 11 dígitos
    if len(numeros) not in [10, 11]:
        return False, "WhatsApp inválido. O número deve ter 10 ou 11 dígitos com o DDD."
    
    # Removida a obrigatoriedade do dígito 9 para aceitar WhatsApp Business (fixo)
    
    return True, ""
    
    # Verificar se o 3º dígito é 9 (celular)
    if numeros[2] != '9':
        return False, "WhatsApp inválido. Deve ser um celular (9º dígito = 9)."
    
    return True, ""


def formatar_telefone(telefone: str) -> str:
    """
    Formata o telefone para (XX) XXXX-XXXX
    
    Args:
        telefone: String com o telefone
    
    Returns:
        String formatada
    
    Exemplo:
        >>> formatar_telefone("6733333333")
        "(67) 3333-3333"
        >>> formatar_telefone("67999999999")
        "(67) 99999-9999"
    """
    numeros = re.sub(r'\D', '', telefone)
    
    if len(numeros) == 10:
        return f"({numeros[0:2]}) {numeros[2:6]}-{numeros[6:10]}"
    elif len(numeros) == 11:
        return f"({numeros[0:2]}) {numeros[2:7]}-{numeros[7:11]}"
    
    return telefone

# ============================================================================
# VALIDADORES DE PLANO
# ============================================================================

PLANOS_VALIDOS = {
    'Basic': {
        'usuarios_min': 1,
        'usuarios_max': 10,
        'valor_mensal': Decimal('50.00'),
        'descricao': 'Acesso básico, 1 admin'
    },
    'PRO': {
        'usuarios_min': 11,
        'usuarios_max': 50,
        'valor_mensal': Decimal('100.00'),
        'descricao': 'Acesso avançado, 3 admins, relatórios'
    },
    'Ultimate': {
        'usuarios_min': 51,
        'usuarios_max': 999,
        'valor_mensal': Decimal('200.00'),
        'descricao': 'Acesso total, suporte 24/7, API'
    }
}

def validar_plano(plano: str) -> Tuple[bool, str]:
    """
    Valida se o plano é válido.
    
    Args:
        plano: String com o nome do plano
    
    Returns:
        Tuple[bool, str]: (é_válido, mensagem_erro)
    """
    
    if plano not in PLANOS_VALIDOS:
        planos = ', '.join(PLANOS_VALIDOS.keys())
        return False, f"Plano inválido. Deve ser um de: {planos}"
    
    return True, ""

def obter_plano_por_usuarios(usuarios: int) -> str:
    """
    Retorna o plano recomendado baseado no número de usuários.
    
    Args:
        usuarios: Número de usuários
    
    Returns:
        String com o nome do plano
    """
    
    for plano, config in PLANOS_VALIDOS.items():
        if config['usuarios_min'] <= usuarios <= config['usuarios_max']:
            return plano
    
    return 'Ultimate'

# ============================================================================
# VALIDADORES DE USUÁRIOS
# ============================================================================

def validar_usuarios(usuarios: int) -> Tuple[bool, str]:
    """
    Valida o número de usuários.
    
    Args:
        usuarios: Número de usuários
    
    Returns:
        Tuple[bool, str]: (é_válido, mensagem_erro)
    """
    
    if not isinstance(usuarios, int):
        return False, "Número de usuários deve ser um inteiro."
    
    if usuarios < 1 or usuarios > 999:
        return False, "Número de usuários deve estar entre 1 e 999."
    
    return True, ""

def arredondar_usuarios(usuarios: int) -> int:
    """
    Arredonda o número de usuários para a dezena superior.
    
    Regra:
    - Se é múltiplo de 10: mantém
    - Se não é: arredonda para próxima dezena
    
    Args:
        usuarios: Número de usuários
    
    Returns:
        Número arredondado
    
    Exemplo:
        >>> arredondar_usuarios(51)
        60
    """
    
    if usuarios % 10 == 0:
        return usuarios
    
    return ((usuarios // 10) + 1) * 10

# ============================================================================
# VALIDADORES DE TREINAMENTO
# ============================================================================

TIPOS_TREINAMENTO = {
    'Online': {
        'setup_padrao': Decimal('0.00'),
        'descricao': 'Sem custo de deslocamento'
    },
    'Presencial': {
        'setup_padrao': Decimal('350.00'),
        'descricao': 'Deslocamento + facilitador'
    },
    'Híbrido': {
        'setup_padrao': Decimal('250.00'),
        'descricao': 'Misto (online + presencial)'
    }
}

def validar_treinamento(treinamento: str) -> Tuple[bool, str]:
    """
    Valida o tipo de treinamento.
    
    Args:
        treinamento: String com o tipo de treinamento
    
    Returns:
        Tuple[bool, str]: (é_válido, mensagem_erro)
    """
    
    if treinamento not in TIPOS_TREINAMENTO:
        tipos = ', '.join(TIPOS_TREINAMENTO.keys())
        return False, f"Tipo de treinamento inválido. Deve ser um de: {tipos}"
    
    return True, ""

def obter_setup_padrao(treinamento: str) -> Decimal:
    """
    Retorna o valor padrão de setup para o tipo de treinamento.
    
    Args:
        treinamento: String com o tipo de treinamento
    
    Returns:
        Decimal com o valor padrão
    """
    
    if treinamento in TIPOS_TREINAMENTO:
        return TIPOS_TREINAMENTO[treinamento]['setup_padrao']
    
    return Decimal('0.00')

# ============================================================================
# VALIDADORES DE VALORES MONETÁRIOS
# ============================================================================

def validar_valor_monetario(valor: float) -> Tuple[bool, str]:
    """
    Valida se um valor monetário é válido.
    
    Args:
        valor: Valor a validar
    
    Returns:
        Tuple[bool, str]: (é_válido, mensagem_erro)
    """
    
    try:
        valor_decimal = Decimal(str(valor))
        
        if valor_decimal < 0:
            return False, "Valor não pode ser negativo."
        
        if valor_decimal > Decimal('999999.99'):
            return False, "Valor muito alto."
        
        return True, ""
    
    except:
        return False, "Valor inválido."

def formatar_moeda(valor: float) -> str:
    """
    Formata um valor em moeda brasileira.
    
    Args:
        valor: Valor a formatar
    
    Returns:
        String formatada (ex: R$ 1.234,56)
    
    Exemplo:
        >>> formatar_moeda(1234.56)
        "R$ 1.234,56"
    """
    
    try:
        valor_decimal = Decimal(str(valor))
        valor_formatado = f"{valor_decimal:,.2f}"
        valor_formatado = valor_formatado.replace(',', 'X')
        valor_formatado = valor_formatado.replace('.', ',')
        valor_formatado = valor_formatado.replace('X', '.')
        return f"R$ {valor_formatado}"
    
    except:
        return f"R$ {valor:.2f}"

def desformatar_moeda(valor_str: str) -> Decimal:
    """
    Converte uma string formatada em moeda para Decimal.
    
    Args:
        valor_str: String formatada (ex: "R$ 1.234,56")
    
    Returns:
        Decimal com o valor
    
    Exemplo:
        >>> desformatar_moeda("R$ 1.234,56")
        Decimal('1234.56')
    """
    
    valor_limpo = valor_str.replace('R$', '').strip()
    valor_limpo = valor_limpo.replace('.', '')
    valor_limpo = valor_limpo.replace(',', '.')
    
    try:
        return Decimal(valor_limpo)
    except:
        return Decimal('0.00')

# ============================================================================
# VALIDADORES DE DESCONTO
# ============================================================================

def validar_desconto(desconto_percentual: float = None, desconto_valor: float = None) -> Tuple[bool, str]:
    """
    Valida o desconto (percentual ou valor).
    
    Args:
        desconto_percentual: Desconto em percentual (0-100)
        desconto_valor: Desconto em valor (R$)
    
    Returns:
        Tuple[bool, str]: (é_válido, mensagem_erro)
    """
    
    if desconto_percentual is not None:
        if desconto_percentual < 0 or desconto_percentual > 100:
            return False, "Desconto percentual deve estar entre 0 e 100%."
        
        if desconto_percentual > 50:
            return False, "Desconto acima de 50% requer aprovação do gerente."
    
    if desconto_valor is not None:
        if desconto_valor < 0:
            return False, "Desconto em valor não pode ser negativo."
    
    return True, ""

def calcular_desconto(valor_total: Decimal, desconto_percentual: float = None, desconto_valor: Decimal = None) -> Decimal:
    """
    Calcula o valor do desconto.
    
    Args:
        valor_total: Valor total da proposta
        desconto_percentual: Desconto em percentual
        desconto_valor: Desconto em valor
    
    Returns:
        Decimal com o valor do desconto
    """
    
    if desconto_percentual is not None:
        return valor_total * Decimal(str(desconto_percentual)) / Decimal('100')
    
    if desconto_valor is not None:
        return Decimal(str(desconto_valor))
    
    return Decimal('0.00')

# ============================================================================
# VALIDADORES DE DATAS
# ============================================================================

def validar_data_inicio_servico(data_str: str) -> Tuple[bool, str]:
    """
    Valida a data de início do serviço.
    
    Args:
        data_str: String com a data (formato: YYYY-MM-DD)
    
    Returns:
        Tuple[bool, str]: (é_válido, mensagem_erro)
    """
    
    try:
        data = datetime.strptime(data_str, '%Y-%m-%d')
        
        if data.date() < datetime.utcnow().date():
            return False, "Data de início não pode ser no passado."
        
        return True, ""
    
    except ValueError:
        return False, "Data inválida. Use o formato YYYY-MM-DD."

def calcular_data_validade(dias: int = 30) -> datetime:
    """
    Calcula a data de validade da proposta.
    
    Args:
        dias: Número de dias de validade (padrão: 30)
    
    Returns:
        datetime com a data de validade
    """
    
    return datetime.utcnow() + timedelta(days=dias)

# ============================================================================
# VALIDAÇÕES CRUZADAS
# ============================================================================

def validar_combinacao_plano_usuarios(plano: str, usuarios: int) -> Tuple[bool, str]:
    """
    Valida a combinação de plano e número de usuários.
    
    Args:
        plano: String com o nome do plano
        usuarios: Número de usuários
    
    Returns:
        Tuple[bool, str]: (é_válido, mensagem_erro)
    """
    
    if plano not in PLANOS_VALIDOS:
        return False, "Plano inválido."
    
    config = PLANOS_VALIDOS[plano]
    
    if usuarios < config['usuarios_min']:
        return False, f"Plano {plano} requer no mínimo {config['usuarios_min']} usuários."
    
    if usuarios > config['usuarios_max']:
        return False, f"Plano {plano} suporta no máximo {config['usuarios_max']} usuários."
    
    return True, ""

# ============================================================================
# FUNÇÃO PRINCIPAL DE VALIDAÇÃO
# ============================================================================

def validar_proposta_completa(dados: Dict[str, Any]) -> Tuple[bool, Dict[str, str]]:
    """
    Valida todos os dados da proposta.
    
    Args:
        dados: Dicionário com os dados da proposta
    
    Returns:
        Tuple[bool, Dict]: (é_válido, dicionário_de_erros)
    """
    
    erros = {}
    
    # Validar CNPJ
    if 'cnpj' in dados:
        valido, msg = validar_cnpj(dados['cnpj'])
        if not valido:
            erros['cnpj'] = msg
    
    # Validar Email
    if 'email' in dados:
        valido, msg = validar_email(dados['email'])
        if not valido:
            erros['email'] = msg
    
    # Validar Telefone
    if 'telefone' in dados and dados['telefone']:
        valido, msg = validar_telefone(dados['telefone'])
        if not valido:
            erros['telefone'] = msg
    
    # Validar WhatsApp
    if 'whatsapp' in dados and dados['whatsapp']:
        valido, msg = validar_whatsapp(dados['whatsapp'])
        if not valido:
            erros['whatsapp'] = msg
    
    # Validar Plano
    if 'plano' in dados:
        valido, msg = validar_plano(dados['plano'])
        if not valido:
            erros['plano'] = msg
    
    # Validar Usuários
    if 'usuarios' in dados:
        valido, msg = validar_usuarios(dados['usuarios'])
        if not valido:
            erros['usuarios'] = msg
    
    # Validar Combinação Plano + Usuários
    if 'plano' in dados and 'usuarios' in dados:
        valido, msg = validar_combinacao_plano_usuarios(dados['plano'], dados['usuarios'])
        if not valido:
            erros['combinacao'] = msg
    
    # Validar Treinamento
    if 'treinamento_tipo' in dados and dados['treinamento_tipo']:
        valido, msg = validar_treinamento(dados['treinamento_tipo'])
        if not valido:
            erros['treinamento_tipo'] = msg
    
    # Validar Desconto
    if 'desconto_percentual' in dados or 'desconto_valor' in dados:
        valido, msg = validar_desconto(
            dados.get('desconto_percentual'),
            dados.get('desconto_valor')
        )
        if not valido:
            erros['desconto'] = msg
    
    return len(erros) == 0, erros

File: app/static/js/validators.js
// app/static/js/validators.js - VERSÃO AJUSTADA PARA SPRINT 1

console.log('✅ validators.js carregado com sucesso');

const DEBUG = typeof window !== 'undefined' ? (window.DEBUG || false) : false;

// ============================================
// VALIDADORES DE CNPJ
// ============================================

function validarCNPJ(cnpj) {
    try {
        const cnpjLimpo = String(cnpj || '').replace(/\D/g, '');

        if (cnpjLimpo.length !== 14) {
            return { valido: false, mensagem: 'CNPJ deve ter 14 dígitos.' };
        }

        if (cnpjLimpo === cnpjLimpo[0].repeat(14)) {
            return { valido: false, mensagem: 'CNPJ inválido (todos os dígitos iguais).' };
        }

        const cnpjBase = cnpjLimpo.substring(0, 12);
        const pesos1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
        let soma = 0;

        for (let i = 0; i < 12; i++) {
            soma += parseInt(cnpjBase[i]) * pesos1[i];
        }

        let resto = soma % 11;
        let dv1 = resto < 2 ? 0 : 11 - resto;

        if (parseInt(cnpjLimpo[12]) !== dv1) {
            return { valido: false, mensagem: 'CNPJ inválido. Dígito verificador incorreto.' };
        }

        const cnpjComDV1 = cnpjBase + dv1;
        const pesos2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
        soma = 0;

        for (let i = 0; i < 13; i++) {
            soma += parseInt(cnpjComDV1[i]) * pesos2[i];
        }

        resto = soma % 11;
        let dv2 = resto < 2 ? 0 : 11 - resto;

        if (parseInt(cnpjLimpo[13]) !== dv2) {
            return { valido: false, mensagem: 'CNPJ inválido. Dígito verificador incorreto.' };
        }

        return { valido: true, mensagem: '' };
    } catch (error) {
        if (DEBUG) console.error('Erro na validação de CNPJ:', error);
        return { valido: false, mensagem: 'Erro ao validar CNPJ.' };
    }
}

// ============================================
// VALIDADORES DE EMAIL
// ============================================

function validarEmail(email) {
    try {
        const emailLimpo = String(email || '').trim();
        const padrao = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

        if (!padrao.test(emailLimpo)) {
            return { valido: false, mensagem: 'Email inválido. Verifique o formato (ex: usuario@exemplo.com).' };
        }

        return { valido: true, mensagem: '' };
    } catch (error) {
        if (DEBUG) console.error('Erro na validação de email:', error);
        return { valido: false, mensagem: 'Erro ao validar email.' };
    }
}

// ============================================
// VALIDADORES DE TELEFONE E WHATSAPP
// ============================================

function validarTelefone(telefone) {
    try {
        const telefoneStr = String(telefone || '').trim();
        const numeros = telefoneStr.replace(/\D/g, '');

        if (numeros.length < 10 || numeros.length > 11) {
            return { valido: false, mensagem: 'Telefone deve ter 10 ou 11 dígitos.' };
        }

        return { valido: true, mensagem: '' };
    } catch (error) {
        if (DEBUG) console.error('Erro na validação de telefone:', error);
        return { valido: false, mensagem: 'Erro ao validar telefone.' };
    }
}

function formatarTelefone(telefone) {
    try {
        const numeros = String(telefone || '').replace(/\D/g, '');

        if (numeros.length === 0) return '';

        if (numeros.length === 10) {
            return `(${numeros.substring(0, 2)}) ${numeros.substring(2, 6)}-${numeros.substring(6, 10)}`;
        }

        if (numeros.length === 11) {
            return `(${numeros.substring(0, 2)}) ${numeros.substring(2, 7)}-${numeros.substring(7, 11)}`;
        }

        return telefone;
    } catch (error) {
        if (DEBUG) console.error('Erro ao formatar telefone:', error);
        return telefone;
    }
}

function validarWhatsApp(whatsapp) {
    try {
        const whatsappStr = String(whatsapp || '').trim();
        const numeros = whatsappStr.replace(/\D/g, '');

        if (numeros.length !== 11 || numeros[2] !== '9') {
            return { valido: false, mensagem: 'WhatsApp inválido. Deve ser um celular com 11 dígitos e 9 após o DDD.' };
        }

        return { valido: true, mensagem: '' };
    } catch (error) {
        if (DEBUG) console.error('Erro na validação de WhatsApp:', error);
        return { valido: false, mensagem: 'Erro ao validar WhatsApp.' };
    }
}

// ============================================
// VALIDADORES DE USUÁRIOS
// ============================================

function validarUsuarios(usuarios) {
    try {
        const num = parseInt(usuarios);

        if (isNaN(num) || num < 1 || num > 999) {
            return { valido: false, mensagem: 'Quantidade de pessoas deve estar entre 1 e 999.' };
        }

        return { valido: true, mensagem: '' };
    } catch (error) {
        if (DEBUG) console.error('Erro na validação de usuários:', error);
        return { valido: false, mensagem: 'Erro ao validar quantidade de pessoas.' };
    }
}

function arredondarUsuarios(usuarios) {
    try {
        const num = parseInt(usuarios);

        if (isNaN(num) || num < 1) return 10;
        if (num % 10 === 0) return num;

        return Math.ceil(num / 10) * 10;
    } catch (error) {
        if (DEBUG) console.error('Erro ao arredondar usuários:', error);
        return 10;
    }
}

// ============================================
// VALIDADORES DE MOEDA
// ============================================

function validarMoeda(valor) {
    try {
        const num = typeof valor === 'string' ? desformatarMoeda(valor) : parseFloat(valor);

        if (isNaN(num) || num < 0) {
            return { valido: false, mensagem: 'Valor não pode ser negativo ou inválido.' };
        }

        if (num > 999999.99) {
            return { valido: false, mensagem: 'Valor muito alto (máximo R$ 999.999,99).' };
        }

        return { valido: true, mensagem: '' };
    } catch (error) {
        if (DEBUG) console.error('Erro na validação de moeda:', error);
        return { valido: false, mensagem: 'Erro ao validar valor monetário.' };
    }
}

function formatarMoeda(valor) {
    try {
        const num = typeof valor === 'string' ? desformatarMoeda(valor) : parseFloat(valor);

        if (isNaN(num)) return 'R$ 0,00';

        return new Intl.NumberFormat('pt-BR', {
            style: 'currency',
            currency: 'BRL'
        }).format(num);
    } catch (error) {
        if (DEBUG) console.error('Erro ao formatar moeda:', error);
        return 'R$ 0,00';
    }
}

function desformatarMoeda(valor) {
    if (!valor || typeof valor !== 'string') return 0;

    let limpo = valor.replace('R$', '').trim();
    limpo = limpo.replace(/\./g, '');
    limpo = limpo.replace(',', '.');

    const num = parseFloat(limpo);
    return isNaN(num) ? 0 : num;
}

// ============================================
// VALIDADORES DE DESCONTO
// ============================================

function validarDesconto(desconto) {
    try {
        const texto = String(desconto || '').trim();

        if (!texto) {
            return { valido: true, mensagem: '', tipo: 'valor', valor: 0 };
        }

        if (texto.includes('%')) {
            const percentual = parseFloat(texto.replace('%', '').replace(',', '.'));

            if (isNaN(percentual) || percentual < 0 || percentual > 100) {
                return { valido: false, mensagem: 'Desconto percentual deve estar entre 0 e 100%.' };
            }

            if (percentual > 50) {
                return { valido: false, mensagem: 'Desconto acima de 50% requer aprovação do gerente.' };
            }

            return { valido: true, mensagem: '', tipo: 'percentual', valor: percentual };
        }

        const valor = desformatarMoeda(texto);

        if (isNaN(valor) || valor < 0) {
            return { valido: false, mensagem: 'Desconto em valor não pode ser negativo ou inválido.' };
        }

        return { valido: true, mensagem: '', tipo: 'valor', valor: valor };
    } catch (error) {
        if (DEBUG) console.error('Erro na validação de desconto:', error);
        return { valido: false, mensagem: 'Erro ao validar desconto.' };
    }
}

// ============================================
// FUNÇÕES DE CÁLCULO
// ============================================

function calcularValorMensal(plano, usuarios) {
    const planosBase = {
        'Basic': 50.00,
        'PRO': 100.00,
        'Ultimate': 200.00,
        'Desktop': 150.00
    };

    const valorBase = planosBase[plano] || 0;
    const multiplicador = Math.max(1, Math.ceil((parseInt(usuarios) || 0) / 10));

    return valorBase * multiplicador;
}

function obterSetupPadrao(treinamento) {
    const setups = {
        'Online': 0.00,
        'Presencial': 350.00,
        'Híbrido': 250.00
    };

    return setups[treinamento] || 0.00;
}

function calcularDesconto(subtotal, descontoInfo) {
    try {
        if (!descontoInfo || !descontoInfo.valido) return 0;

        if (descontoInfo.tipo === 'percentual') {
            return subtotal * (descontoInfo.valor / 100);
        }

        return descontoInfo.valor;
    } catch (error) {
        if (DEBUG) console.error('Erro ao calcular desconto:', error);
        return 0;
    }
}

function calcularTotal(valorMensal, setup, treinamento, desconto, servicosAdicionais = 0, taxaUnica = 0) {
    try {
        const subtotal = valorMensal + setup + treinamento + servicosAdicionais + taxaUnica;
        const total = Math.max(subtotal - desconto, 0);

        return { subtotal, total };
    } catch (error) {
        if (DEBUG) console.error('Erro ao calcular total:', error);
        return { subtotal: 0, total: 0 };
    }
}

// ============================================
// SERVIÇOS ADICIONAIS
// ============================================

function validarServicosAdicionais(servicos, usuarios, plano) {
    try {
        const faixa = arredondarUsuarios(usuarios).toString();

        const tabelaPrecos = {
            'licenca_facial': { '10': 20, '20': 35, '30': 50, '40': 65, '50': 80, '60': 95 },
            'gestao_arquivos': { '10': 15, '20': 25, '30': 35, '40': 45, '50': 55, '60': 65 },
            'controle_ferias': { '10': 10, '20': 18, '30': 26, '40': 34, '50': 42, '60': 50 },
            'requis_calc_int': { '10': 25, '20': 45, '30': 65, '40': 85, '50': 105, '60': 125 }
        };

        let totalAdicionais = 0;
        let mensagem = '';

        Object.keys(servicos || {}).forEach(servico => {
            const qtd = parseInt(servicos[servico]) || 0;
            if (qtd > 0) {
                const precoFaixa = tabelaPrecos[servico] ? tabelaPrecos[servico][faixa] || 0 : 0;
                if (precoFaixa === 0) {
                    mensagem += `Preço não definido para ${servico} na faixa ${faixa}. `;
                } else {
                    totalAdicionais += qtd * precoFaixa;
                }
            }
        });

        return { valido: mensagem === '', mensagem, total: totalAdicionais };
    } catch (error) {
        if (DEBUG) console.error('Erro na validação de serviços adicionais:', error);
        return { valido: false, mensagem: 'Erro ao validar serviços adicionais.', total: 0 };
    }
}

// ============================================
// DEMAIS FUNÇÕES
// ============================================

function calcularDataValidade(dias = 10) {
    try {
        const hoje = new Date();
        hoje.setDate(hoje.getDate() + dias);
        return hoje.toISOString().split('T')[0];
    } catch (error) {
        if (DEBUG) console.error('Erro ao calcular data de validade:', error);
        return new Date().toISOString().split('T')[0];
    }
}

function validarFormularioCompleto(formData) {
    try {
        const erros = {};

        const cnpjResult = validarCNPJ(formData.cnpj || '');
        if (!cnpjResult.valido) erros.cnpj = cnpjResult.mensagem;

        const emailResult = validarEmail(formData.email || '');
        if (!emailResult.valido) erros.email = emailResult.mensagem;

        if (formData.telefone) {
            const telResult = validarTelefone(formData.telefone);
            if (!telResult.valido) erros.telefone = telResult.mensagem;
        }

        const whatsappResult = validarWhatsApp(formData.whatsapp || '');
        if (!whatsappResult.valido) erros.whatsapp = whatsappResult.mensagem;

        const usuariosResult = validarUsuarios(formData.usuarios || 0);
        if (!usuariosResult.valido) erros.usuarios = usuariosResult.mensagem;

        ['setup', 'treinamentoValor', 'taxaUnica'].forEach(campo => {
            if (formData[campo]) {
                const moedaResult = validarMoeda(formData[campo]);
                if (!moedaResult.valido) erros[campo] = moedaResult.mensagem;
            }
        });

        const descontoResult = validarDesconto(formData.desconto || '');
        if (!descontoResult.valido) erros.desconto = descontoResult.mensagem;

        if (formData.servicosAdicionais) {
            const servicosResult = validarServicosAdicionais(formData.servicosAdicionais, formData.usuarios || 0, formData.plano || '');
            if (!servicosResult.valido) erros.servicos = servicosResult.mensagem;
        }

        const valido = Object.keys(erros).length === 0;

        let total = 0;
        if (valido) {
            const calc = calcularTotal(
                calcularValorMensal(formData.plano || '', formData.usuarios || 0),
                desformatarMoeda(formData.setup || '0'),
                desformatarMoeda(formData.treinamentoValor || '0'),
                calcularDesconto(
                    calcularValorMensal(formData.plano || '', formData.usuarios || 0) +
                    desformatarMoeda(formData.setup || '0') +
                    desformatarMoeda(formData.treinamentoValor || '0'),
                    descontoResult
                ),
                formData.servicosTotal || 0,
                desformatarMoeda(formData.taxaUnica || '0')
            );
            total = calc.total;
        }

        return { valido, erros, total, dataValidade: calcularDataValidade(10) };
    } catch (error) {
        if (DEBUG) console.error('Erro na validação geral do formulário:', error);
        return { valido: false, erros: { geral: 'Erro interno na validação.' }, total: 0 };
    }
}

async function consultarCNPJ(cnpj) {
    try {
        const cnpjLimpo = String(cnpj || '').replace(/\D/g, '');

        if (cnpjLimpo.length !== 14) {
            return { sucesso: false, mensagem: 'CNPJ deve ter 14 dígitos.' };
        }

        const resposta = await fetch(`https://brasilapi.com.br/api/cnpj/v1/${cnpjLimpo}`);

        if (!resposta.ok) {
            return { sucesso: false, mensagem: 'CNPJ não encontrado na consulta.' };
        }

        const dados = await resposta.json();

        return {
            sucesso: true,
            razaoSocial: dados.razao_social || '',
            nomeFantasia: dados.nome_fantasia || '',
            cnpj: dados.cnpj || cnpjLimpo
        };
    } catch (error) {
        return { sucesso: false, mensagem: 'Erro ao consultar CNPJ.' };
    }
}

console.log('✅ validators.js finalizado - Todas as funções disponíveis.');